### PR TITLE
Implement LWG-3520

### LINKED ISSUE #2180: LWG-3520 `iter_move` and `iter_swap` are inconsistent for `transform_view::iterator`
LWG-3520 `iter_move` and `iter_swap` are inconsistent for `transform_view::`*`iterator`*

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #2180

### CODE REVIEW

**Review by miscco — COMMENTED** [review]
Looks good.

To ensure that we indeed cannot do this, it would be great if you would add a concept ``CanIterSwap` to range_algorithm_support.h like
```cpp
template <class I>
concept CanIterSwap = requires(I&& i1, I&& i2) {
    ranges::iter_swap(std::forward<I>(i1), std::forward<I>(i2));
};
```

and verify that it is not fulfilled

File: tests/std/tests/P0896R4_views_transform/test.cpp [hunk 1]
```diff
@@ -509,16 +509,8 @@ struct iterator_instantiator {
...
-                auto i1 = ranges::next(i0);
-                ranges::iter_swap(i0, i1);
-                assert(mutable_ints[0] == 1);
-                assert(mutable_ints[1] == 0);
-                ranges::iter_swap(i1, i0);
-                assert(mutable_ints[0] == 0);
-                assert(mutable_ints[1] == 1);
-                STATIC_ASSERT(NOEXCEPT_IDL0(ranges::iter_swap(i0, i1)));
-            }
+            // LWG-3520 forbids iter_swap for transform_view::iterator
```
Comment (AdamBucior) [inline tests/std/tests/P0896R4_views_transform/test.cpp:512]: This comment is misleading. Calling `iter_swap` on `transform_view::iterator` is not forbidden, it's just impossible when the result of transformation is not a lvalue reference. We should also add test coverage for that.

**Review by CaseyCarter — APPROVED** [review]
Looks great.

File: tests/std/tests/P0896R4_views_transform/test.cpp [hunk 2]
```diff
@@ -784,4 +775,19 @@ int main() {
 
         (void) views::transform(Fn{})(span<int>{});
     }
+
+    { // Validate that iter_swap works when result of transformation is a lvalue reference
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_transform/test.cpp:779]: Grammar corner case: This should say "an" because even though "lvalue" starts with a consonant, it's pronounced as the letter "L" followed by "value", and when spoken, "el" starts with a vowel sound.

```suggestion
    { // Validate that iter_swap works when result of transformation is an lvalue reference
```

Existing examples:

https://github.com/microsoft/STL/blob/d6f9987d7ec694b8599fe21e87e3653c36566223/stl/inc/type_traits#L254
https://github.com/microsoft/STL/blob/d6f9987d7ec694b8599fe21e87e3653c36566223/stl/inc/type_traits#L263

Other examples of letters with consonant and vowel sounds (here, the first letter controls "a" vs. "an"):

* "My old computer had a CRT." ("a see are tee")
* "My new computer has an SSD." ("an ess ess dee")
* "This Mars rover has an RTG." ("an are tee gee")

**Review by StephanTLavavej — COMMENTED** [review]
Looks good, thanks! I'll push a comment change.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to an MSVC-internal PR. Please notify me if any further changes are pushed.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for implementing and testing this LWG issue resolution, one step closer to 100% conformance! :100: :smile_cat: :heavy_check_mark:
